### DISCUSSION TITLE
Using SQLAlchemy Events for validation

### DISCUSSION THREAD

**Discussion by zikphil** [sqlalchemy/sqlalchemy#5971] [Usage Questions] [answered]
Hi amazing, god-tier community,

So I want to use some before_insert and before_update ORM Event Hooks in order to perform validation before saving an object.

i.e:

- User changes Task Owner from ID 2 to ID 3
- ORM Event Hook runs:
-      --> Detects that User ID 3 already has this task so instead of saving the task object, it deletes it.

Stuff like that. Since I have a common shared library between my Celery Worker and my Async API Backend I figured using those hooks would be an ideal way to ensure that I can apply some global validation rules no matter where the change originated from.

Are there any red flags by doing it this way? Will I be getting some Session errors when attempting to delete an ORM object during it's own hooks? Are event hooks the right way to go about it?

Any wisdom would be appreciated.

Have an heavenly day.

**Comment by zzzeek** [comment, marked as answer]
if your validations are doing things involving the relationships of objects then it's better to use the [before_flush](https://docs.sqlalchemy.org/en/13/orm/events.html?highlight=before_flush#sqlalchemy.orm.events.SessionEvents.before_flush) hook, because in there, you can manipulate the collections and related attributes freely.      if your validations are involving strictly the local column-bound attributes and not using relationships, then the before_insert() / before_update() is OK.
